### ISSUE TITLE
No success... suggestions?

### ISSUE THREAD

**Issue by danieleghisi** [phreeza/keras-GAN#2]
Hi,

After having updated the code to work with latest keras, I have been testing it for a while, but I'm having no success whatsoever. After a large number of epochs, I'm still sticking with generated white noise.

I'm no expert in GANs... is there anyone having success with these models?
If yes, with which settings/audio files?

Thanks,
Daniele

**Comment by phreeza** [comment]
Hi Daniele,

I have not had any great success with this approach either, in the audio domain I am only getting some sort of vaguely harmonic sounding 'soundscapes', nothing sounding really like music. I have some other work with RNNs, which sound a bit more promising here: https://github.com/phreeza/ml-music/tree/master/tensorflow maybe you want to play with that.

Also, do you maybe want to send me a PR for the updates you made to the latest Keras?

Thomas

**Comment by phreeza** [comment]
Actually my latest experiments are in this branch: https://github.com/phreeza/tensorflow-vrnn/tree/music so check that out if you are interested.

**Comment by danieleghisi** [comment]
Hi, and thanks for answering.

I'd send a PR, but I'm not sure about some things (e.g. set_state() and get_state() for the optimizers seem to have disappeared in recent keras versions, but I have no idea about what replaces them). I just deleted them, but that was probably not a smart choice. The other thing that didn't work any longer was r.totals['loss']/batch_size; I've replaced all instances with r.history['loss'][0], and this seems to work properly. 

Thanks for pointing out to your other experiments. As far as GANs are concerned, I'm having a bit more luck with this https://github.com/soumith/dcgan.torch applied to spectrograms, but nothing astonishing yet. I'll definitely check out your vrnn!

**Comment by phreeza** [comment]
OK if you are looking at spectrograms, maybe also try it on the MDCT spectrograms which I am using in the branch I posted above (in utils.py) They behave very similar to FFT spectrograms but can preserve the phase information and reduce glitches between frames. (used in OGG compression for example).

**Comment by danieleghisi** [comment]
Thanks, I didn't know this phase property of MDCTs - well, ok, since MDCTs are real-valued, I guess I should have thought of that... I'll definitely have a look.

For the time being I'm using this extremely simple iterative algorithm to reconstruct phases in STFTs:
http://dsp.stackexchange.com/questions/9877/reconstruction-of-audio-signal-from-spectrogram/13401#13401
Although rather slow, results are quite good...

**Comment by engharat** [comment]
Unfortunately, because of this nasty bug [https://github.com/fchollet/keras/issues/4674](url) GAN models cannot work on keras: trainable  flag is not working!

**Comment by phreeza** [comment]
@engharat from the discussion in the issue you linked to it looks like this is not really a bug, are you sure it affects this code? Were you able to solve the problem?

**Comment by phreeza** [comment]
If you do still have an issue, please make a new  issue for that. closing for now.

**Comment by engharat** [comment]
As you can read from that discussion, the actual situation is not clear. Beside calling it a "bug" or an "issue", the fact that batch normalization is changing the whole network output while training only a part of the network is a problem from my point of view. For example, when you train the generator feeding the input on the whole GAN network (keeping discriminator frozen) the discriminator is still changing because of its batchnorm. This could be or could not be a problem, depending on your GAN application. One of the solution, proposed into the discussion, have been to put batchnorm momentum = 1. I'm starting to hate how BN behave in GAN models, so I moved away from it, building GANs without BN (using my custom weight normalization implementation).

**Comment by phreeza** [comment]
Ah ok, thanks for clarifying. So as long as we don't use BN we should be fine?

**Comment by engharat** [comment]
It seems so. If you make a network C that contains subnetwork B with Conv&BN layers and a subnetwork A(the GAN case) training A via C while keeping B frozen results in B updating its BN parameters, and so B output change while it shouldn't. Still I don't know (and it isn't clear) how much this behavior affect overall GAN performance. 
In the previous example, if network B doesn't have any BN layer, there isn't any issue. 
Finally,  depending on your GAN task, BN could improve or worsening your final results, and often it is safe to remove it(increasing the number of iteration before convergency)
